editor/components/colors/with-colors.js

```diff
@@ -1,91 +1,119 @@
+	if ( isFunction( args[ 0 ] ) ) {
+		deprecated( 'Using withColors( mapGetSetColorToProps ) ', {
+			version: '3.2',
+			alternative: 'withColors( colorAttributeName, { secondColorAttributeName: \'color-context\' }, ... )',
+		} );
+		return withColorsDeprecated( args[ 0 ] );
+	}
+
+	const colorMap = reduce( args, ( colorObj, arg ) => {
+		return {
+			...colorObj,
+			...( isString( arg ) ? { [ arg ]: kebabCase( arg ) } : arg ),
+		};
+	}, {} );
+
+	const relevantAttributeNames = flatMap(
+		colorMap,
+		( value, attributeName ) => [ attributeName, `custom${ upperFirst( attributeName ) }` ]
+	);
+
+	return createHigherOrderComponent(
+		compose( [
+			withSelect( ( select ) => {
 				const settings = select( 'core/editor' ).getEditorSettings();
 				return {
 					colors: get( settings, [ 'colors' ], DEFAULT_COLORS ),
 				};
 			} ),
-		( WrappedComponent ) => {
-			return class extends Component {
-				constructor() {
-					super( ...arguments );
-					/**
-					* Even though, we don't expect setAttributes or colors to change memoizing it is essential.
-					* If setAttributes or colors are not memoized, each time memoizedGetColor/memoizedSetColor are called:
-					* a new function reference is returned (even if arguments have not changed).
-					* This would make our memoized chain useless.
-					*/
-					this.memoizedGetColor = memoize( this.memoizedGetColor, { maxSize: 1 } );
-					this.memoizedSetColor = memoize( this.memoizedSetColor, { maxSize: 1 } );
-				}
+			( WrappedComponent ) => {
+				return class extends Component {
+					constructor( props ) {
+						super( props );
+						this.state = {};
+					}
 
-				memoizedGetColor( colors ) {
-					return memoize(
-						( colorName, customColorValue, colorContext ) => {
-							return {
+					static getDerivedStateFromProps( { attributes, colors, setAttributes }, prevState ) {
+						const relevantAttributes = pick( attributes, relevantAttributeNames );
+						if (
+							colors === prevState.colors &&
+							setAttributes === prevState.setAttributes &&
+							isShallowEqual( relevantAttributes, prevState.relevantAttributes )
+						) {
+							return null;
+						}
+
+						const mergeProps = reduce( colorMap, ( mergePropsAcc, colorContext, colorAttributeName ) => {
+							const ufColorAttrName = upperFirst( colorAttributeName );
+							const customColorAttributeName = `custom${ ufColorAttrName }`;
+							const colorName = attributes[ colorAttributeName ];
+							const customColor = attributes[ customColorAttributeName ];
+							mergePropsAcc[ colorAttributeName ] = {
 								name: colorName,
 								class: getColorClass( colorContext, colorName ),
-								value: getColorValue( colors, colorName, customColorValue ),
+								value: getColorValue( colors, colorName, customColor ),
 							};
-						}
-					);
-				}
+							mergePropsAcc[ `set${ ufColorAttrName }` ] =
+								setColorValue( colors, colorAttributeName, customColorAttributeName, setAttributes );
```

Comment:
Rather than copying `setAttributes` from props into state, could instead we assign the prop callback as an instance-bound function?

Similar to what we do in `withDispatch`:

https://github.com/WordPress/gutenberg/blob/eb4fc18fa1c8cae5daf22afd51b9fe22c7f89513/packages/data/src/index.js#L384-L393

Getting a general sense we might be trying to hard to optimize this with caching copies of `relevantAttributes`, `colors`, `setAttributes`. Maybe we could compare against `prevState.mergeProps` to see if the attribute value has changed, rather than doing the shallow equality before.

Something like:

```jsx
return class extends Component {
	constructor( props ) {
		super( props );
		this.state = {};
	}

	static getDerivedStateFromProps( { attributes, colors, setAttributes }, prevState ) {
		return reduce( colorMap, ( mergePropsAcc, colorContext, colorAttributeName ) => {
			const ufColorAttrName = upperFirst( colorAttributeName );
			const customColorAttributeName = `custom${ ufColorAttrName }`;
			const colorName = attributes[ colorAttributeName ];
			const customColor = attributes[ customColorAttributeName ];
			const colorValue = getColorValue( colors, colorName, customColor );
			if ( get( mergePropsAcc, [ colorAttributeName, 'value' ] ) === colorValue ) {
				return mergePropsAcc;
			}

			mergePropsAcc[ colorAttributeName ] = {
				name: colorName,
				class: getColorClass( colorContext, colorName ),
				value: colorValue,
			};
			mergePropsAcc[ `set${ ufColorAttrName }` ] =
				setColorValue( colors, colorAttributeName, customColorAttributeName, setAttributes );
			return mergePropsAcc;
		}, prevState );
	}

	render() {
		return (
			<WrappedComponent
				{ ...{
					...this.props,
					colors: undefined,
					...this.state,
				} }
			/>
		);
	}
};
```

Reply:
Hi @aduth, I changed the code in the direction of your feedback.